sycl/test-e2e/KernelAndProgram/free_function_kernels.cpp

```diff
@@ -0,0 +1,274 @@
+  queue &Queue;
+  std::vector<kernel_id> AllKernelIDs;
+
+public:
+  KernelFinder(queue &Q) : Queue(Q) {
+    // Obtain kernel bundle
+    kernel_bundle Bundle =
+        get_kernel_bundle<bundle_state::executable>(Queue.get_context());
+    std::cout << "Bundle obtained\n";
+    AllKernelIDs = get_kernel_ids();
+    std::cout << "Number of kernels = " << AllKernelIDs.size() << std::endl;
+    for (auto K : AllKernelIDs) {
+      std::cout << "Kernel obtained: " << K.get_name() << std::endl;
+    }
+    std::cout << std::endl;
+  }
+
+  kernel get_kernel(const char *name) {
+    kernel_bundle Bundle =
+        get_kernel_bundle<bundle_state::executable>(Queue.get_context());
+    for (auto K : AllKernelIDs) {
+      auto Kname = K.get_name();
+      if (strcmp(name, Kname) == 0) {
+        kernel Kernel = Bundle.get_kernel(K);
+        return Kernel;
+      }
+    }
+    std::cout << "No kernel named " << name << " found\n";
+    exit(1);
+  }
+};
+
+void printUSM(int *usmPtr, int size) {
+  std::cout << "usmPtr[] = {";
+  for (int i = 0; i < size; i++) {
+    std::cout << usmPtr[i] << ", ";
+  }
+  std::cout << "}\n";
+}
+
+bool checkUSM(int *usmPtr, int size, int *Result) {
+  bool Pass = true;
+  for (int i = 0; i < size; i++) {
+    if (usmPtr[i] != Result[i]) {
+      Pass = false;
+      break;
+    }
+  }
+  if (Pass)
+    return true;
+
+  std::cout << "Expected = {";
+  for (int i = 0; i < size; i++) {
+    std::cout << Result[i] << ", ";
+  }
+  std::cout << "}\n";
+  std::cout << "Result   = {";
+  for (int i = 0; i < size; i++) {
+    std::cout << usmPtr[i] << ", ";
+  }
+  std::cout << "}\n";
+  return false;
+}
+
+SYCL_EXTERNAL
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(
+    (ext::oneapi::experimental::single_task_kernel))
+void ff_0(int *ptr, int start, int end) {
+  for (int i = start; i <= end; i++)
+    ptr[i] = start + end;
+}
+
+bool test_0(queue Queue, KernelFinder &KF) {
+  constexpr int Range = 10;
+  int *usmPtr = malloc_shared<int>(Range, Queue);
+  int start = 3;
+  int end = 5;
+  int Result[Range] = {0, 0, 0, 8, 8, 8, 0, 0, 0, 0};
+  range<1> R1{Range};
+
+  memset(usmPtr, 0, Range * sizeof(int));
+  Queue.submit([&](handler &Handler) {
+    Handler.single_task([=]() {
+      for (int i = start; i <= end; i++)
+        usmPtr[i] = start + end;
+    });
+  });
+  Queue.wait();
+  bool PassA = checkUSM(usmPtr, Range, Result);
+  std::cout << "Test 0a: " << (PassA ? "PASS" : "FAIL") << std::endl;
+
+  kernel Kernel = KF.get_kernel("_Z18__sycl_kernel_ff_0Piii");
+  memset(usmPtr, 0, Range * sizeof(int));
+  Queue.submit([&](handler &Handler) {
+    Handler.set_arg(0, usmPtr);
+    Handler.set_arg(1, start);
+    Handler.set_arg(2, end);
+    Handler.single_task(Kernel);
+  });
+  Queue.wait();
+  bool PassB = checkUSM(usmPtr, Range, Result);
+  std::cout << "Test 0b: " << (PassB ? "PASS" : "FAIL") << std::endl;
+
+  free(usmPtr, Queue);
+  return PassA && PassB;
+}
+
+// Overloaded free function definition.
+SYCL_EXTERNAL
```

Comment:
@rdeodhar  - I thought I saw this asked by someone else, but I don't see it now.

I notice that if SYCL_EXTERNAL is not used that it seems the kernel doesn't get output.  Is this always going to be required?  

I believe they want to support having these in namespaces, where we won't be able to use SYCL_EXTERNAL

Reply:
This is a temporary limitation. See comment about joint_matrix headers causing problems in scanning the SYCLAddIRAttributes.

Reply:
The spec requires that free functions be at namespace scope or static members of a class.

Reply:
I expect that DPC++ will only support free function kernels that are defined at namespace scope.  See this section of the spec:

https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_free_function_kernels.asciidoc#restrictions-for-integration-header-implementations

We could relax this restriction later when we implement integration without using the integration header.